extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java

```diff
@@ -164,34 +168,88 @@ public QueryDefinition makeQueryDefinition(
+    final ShuffleSpec nextShuffleWindowSpec;
+    if (originalQuery.getContext().containsKey(MultiStageQueryContext.NEXT_WINDOW_SHUFFLE_COL)) {
+      final ClusterBy windowClusterBy = (ClusterBy) originalQuery.getContext().get(MultiStageQueryContext.NEXT_WINDOW_SHUFFLE_COL);
+      nextShuffleWindowSpec = new HashShuffleSpec(
+          windowClusterBy,
+          maxWorkerCount
+      );
+    } else {
+      nextShuffleWindowSpec = null;
+    }
+    final ShuffleSpec stageShuffleSpec;
+    if (shuffleSpecFactoryPostAggregation != null) {
+      List<KeyColumn> columns = resultClusterBy.getColumns();
+      if (nextShuffleWindowSpec != null) {
+        columns.addAll(nextShuffleWindowSpec.clusterBy().getColumns());
+        // Creating a new cluster by with the columns from existing
+        // plus the columns from the next window partition column
+        final ClusterBy tmp = new ClusterBy(columns, resultClusterBy.getBucketByCount());
+        stageShuffleSpec = shuffleSpecFactoryPostAggregation.build(tmp, false);
+      } else {
+        stageShuffleSpec = shuffleSpecFactoryPostAggregation.build(resultClusterBy, false);
+      }
+    } else {
+      stageShuffleSpec = nextShuffleWindowSpec;
+    }
+    final RowSignature stageSignature;
+    if (stageShuffleSpec == null) {
+      stageSignature = resultSignature;
+    } else {
+      // sort the signature to make sure the prefix is aligned
+      stageSignature = QueryKitUtils.sortableSignature(
+          resultSignature,
+          stageShuffleSpec.clusterBy().getColumns()
+      );
+    }
 
     if (doLimitOrOffset) {
+      queryDefBuilder.add(
+          StageDefinition.builder(firstStageNumber + 1)
+                         .inputs(new StageInputSpec(firstStageNumber))
+                         .signature(resultSignature)
+                         .maxWorkerCount(maxWorkerCount)
+                         .shuffleSpec(
+                             shuffleSpecFactoryPostAggregation != null
+                             ? shuffleSpecFactoryPostAggregation.build(resultClusterBy, false)
+                             : null
+                         )
+                         .processorFactory(new GroupByPostShuffleFrameProcessorFactory(queryToRun))
+      );
       final DefaultLimitSpec limitSpec = (DefaultLimitSpec) queryToRun.getLimitSpec();
       queryDefBuilder.add(
           StageDefinition.builder(firstStageNumber + 2)
                          .inputs(new StageInputSpec(firstStageNumber + 1))
                          .signature(resultSignature)
                          .maxWorkerCount(1)
-                         .shuffleSpec(null) // no shuffling should be required after a limit processor.
+                         // no shuffling should be required after a limit processor.
+                         // but need one if the next stage is a window with a partition by
+                         .shuffleSpec(nextShuffleWindowSpec)
```

Comment:
We are not sure if windowing is there. 
Currently the code is a little obtuse and assuming that if window shuffle spec is null, no windowing is there. 
I feel for readability we should make the nextShuffleWindowSpec optional and then do a get here.

Reply:
Refactored query kit to make sure the old code runs when the next window shuffle soec in null else go through the new code